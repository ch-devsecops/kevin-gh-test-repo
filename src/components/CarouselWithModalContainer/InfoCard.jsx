import React from 'react';
import PropTypes, { oneOfType } from 'prop-types';
import { useTranslation } from 'react-i18next';
import { Box, Fade, Icon, Image, Link, Markdown, MarkdownHeading } from '@honda-canada/design-system-react';
import { InView } from 'react-intersection-observer';
import { useSitecoreContext } from '@sitecore-jss/sitecore-jss-react';
import themeStyles from './InfoCard.styles';
import { getIsDarkMode, mapFieldToDesignSystemImage, mapGTMCategory } from '../../utils/sitecoreFields';
import { JSSFieldPropType } from '../../utils/propTypes';
import { getGtmTagValue } from '../../utils/gtmEvents';
import { getBackgroundColorToken, getForegroundColorToken } from '../../utils/carouselSlider';

const Container = themeStyles.apply(Box, 'InfoContainer');
const Heading = themeStyles.apply(MarkdownHeading, 'Heading');
const TextContainer = themeStyles.apply(Box, 'TextContainer');
const BodyText = themeStyles.apply(Markdown, 'BodyText');

const InfoCard = ({ fields, params, rendering, setIndex, ...rest }) => {
  const { t } = useTranslation();
  const { sitecoreContext } = useSitecoreContext() || {};

  if (!fields) return null;

  const { mediaImage, bodyText, title, gtmCategory } = fields;
  const image = mapFieldToDesignSystemImage(mediaImage, undefined, { objectFit: 'contain' });
  const gtmTags = {
    type: rendering?.componentName,
    category: mapGTMCategory(gtmCategory),
  };
  const isDark = getIsDarkMode(sitecoreContext?.route);
  const cardBackgroundColorToken = params?.cardBgColour?.toLowerCase();
  const cardBackgroundColor = getBackgroundColorToken(isDark, cardBackgroundColorToken);
  const cardForegroundColor = getForegroundColorToken(isDark, cardBackgroundColorToken);

  return (
    <InView triggerOnce>
      {({ inView, ref }) => (
        <Fade ref={ref} shouldAnimate={inView} height="100%" initialOpacity={0}>
          <Container
            color={cardForegroundColor}
            data-gtm-component-type={getGtmTagValue(gtmTags?.type)}
            data-gtm-category={getGtmTagValue(gtmTags?.category)}
            {...rest}
          >
            <Box>
              <Image src={image?.props?.src} alt={image?.props?.alt} />
            </Box>
            <TextContainer backgroundColor={cardBackgroundColor} color={cardForegroundColor}>
              <Heading headingOverride="h6" color="inherit">
                {title?.value}
              </Heading>
              <BodyText color="inherit">{bodyText?.value}</BodyText>
              <Box justifyContent="center">
                {/* eslint-disable-next-line jsx-a11y/anchor-is-valid */}
                <Link as="button" styling="primary" onClick={setIndex} color={`${cardForegroundColor} !important`}>
                  {t('Shared.Common.learnMoreLabel')}
                  <Icon name="arrowRight" ml="xs" height="9px" width="5px" iconColor="primary" />
                </Link>
              </Box>
            </TextContainer>
          </Container>
        </Fade>
      )}
    </InView>
  );
};

InfoCard.propTypes = {
  fields: PropTypes.shape({
    mediaImage: PropTypes.shape({
      value: oneOfType([
        PropTypes.string,
        PropTypes.shape({
          alt: PropTypes.string,
          height: PropTypes.string,
          src: PropTypes.string,
          width: PropTypes.string,
        }),
      ]),
    }),
    modalMediaImage: PropTypes.shape({
      value: oneOfType([
        PropTypes.string,
        PropTypes.shape({
          alt: PropTypes.string,
          height: PropTypes.string,
          src: PropTypes.string,
          width: PropTypes.string,
        }),
      ]),
    }),
    title: JSSFieldPropType,
    bodyText: JSSFieldPropType,
    videoUrl: PropTypes.shape({
      src: PropTypes.string,
      ariaLabel: PropTypes.string,
      closeAriaLabel: PropTypes.string,
      onPlay: PropTypes.func,
    }),
    modalTitle: JSSFieldPropType,
    modalBodyText: JSSFieldPropType,
    gtmCategory: JSSFieldPropType,
  }),
  params: PropTypes.shape({
    autoId: PropTypes.string,
    cardAlignment: PropTypes.string,
    cardBgColour: PropTypes.string,
    cardHasPadding: PropTypes.string,
    cardsPerPage: PropTypes.string,
    containerBgColour: PropTypes.string,
    containerBottomMargin: PropTypes.string,
    containerTopMargin: PropTypes.string,
  }),
  rendering: PropTypes.shape({
    componentName: PropTypes.string,
  }),
  setIndex: PropTypes.func,
};

export default InfoCard;
