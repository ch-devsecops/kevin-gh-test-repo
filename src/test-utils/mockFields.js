/* eslint-disable max-len */
/* eslint-disable max-classes-per-file */

// FIXME: jsdoc popup not working on test files
/** Class representing a mock data. */
class MockData {
  /**
   * Add a field.
   * @param {string} key - key of the field obj: i.e. 'title'
   * @param {Object} value - value of the field obj: i.e. { value: 'some text' }.
   */
  add(key, value) {
    this[key] = value;
  }
}

/**
 * Class representing mock fields object.
 * @extends MockData
 *
 * available method: add()
 *
 * default fields: [anchorId, title, bodyText, image, ctaLink, ctaType, ctaIcon, gtmCategory, gtmTitle, contentTitle, subtitle]
 *
 * note: we can add more commonly used fields in the class declaration if needed
 */
class MockFields extends MockData {
  constructor() {
    super();

    this.title = { value: 'Title' };
    this.bodyText = { value: 'BodyText' };
    this.image = {
      value: { src: 'mock.png', alt: 'mock alt' },
    };
    this.ctaLink = {
      value: {
        href: 'www.source.ca',
        title: 'cta aria label',
        text: 'CTA Text',
      },
    };
    this.ctaType = {
      value: 'Tertiary',
    };
    this.ctaIcon = {
      fields: {
        value: {
          value: 'arrowRight',
        },
      },
    };
    this.gtmCategory = {
      fields: {
        value: {
          value: 'Others',
        },
      },
    };
    this.gtmTitle = {
      value: 'gtm title',
    };
    this.contentTitle = {
      value: 'Content Title',
    };
    this.subtitle = {
      value: 'Content Subtitle',
    };
    this.seoH1 = {
      value: '',
    };
    this.anchorId = {
      value: 'anchor-id',
    };
    this.ctaLabel = {
      value: 'CTA Label',
    };
  }
}

/**
 * Class representing a mock rendering object
 * @extends MockData
 *
 * available method: add()
 */
class MockRendering extends MockData {
  /**
   * Create a mock rendering object
   * @param {string} componentName - adds componentName
   */
  constructor(componentName) {
    super();

    this.componentName = componentName;
  }
}

export { MockFields, MockRendering };
