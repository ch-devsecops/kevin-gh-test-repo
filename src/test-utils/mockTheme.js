/**
 * Mock styled-components theme, if you want to test how your component
 * implements theme styles. Import as-is, or spread in any
 * additional values you need.
 *
 * Example usage:

  it('should apply the bgColor param', () => {
    const { colors } = MockTheme;
    const { container } = render(
      <MyComponent
        fields={mockFields}
        params={{ bgColor: 'Gray' }}
      />,
    );
    const wrapper = container.querySelector('section');

    expect(wrapper).toHaveStyleRule('background-color', colors.grey[5]);
  });
 */

import { HONDA_THEME_NAME } from '../utils/constants';

const mockTheme = {
  letterSpacings: {
    sm: '0.5px',
    md: '1px',
  },
  lineHeights: {
    sm: '22px',
    md: '26px',
  },
  colors: {
    white: '#fff',
    transparent: 'transparent',
    grey: ['#707070', '#C7C7C7', '#D5D5D5', '#EAEAEA', '#F8F8F8', '#F5F5F5'],
    black: '#2C2C2C',
    red: '#DA2C2E',
    redShades: ['rgba(200, 16, 46, 0.1)'],
    aodaRed: '#D92629',
    lightRed: '#EC9597',
    darkRed: '#A21C1D',
    blue: '#334C96',
    darkBlue: '#002D62',
    lightBlue: '#00447826',
    yellow: ['#F9CA52', '#EDAB00', '#8D763F'],
    fadedBlue: '#ADB7D5',
    primary: '#DA2C2E',
    primaryAoda: '#D92629',
    primaryDark: '#A21C1D',
    primaryLight: '#EC9597',
    secondary: '#2C2C2C',
    secondaryLight: '#C7C7C7',
    tertiary: '#002D62',
    tertiaryLight: '#334C96',
    tertiaryDisabled: '#ADB7D5',
    danger: '#DA2C2E',
    warning: '#2C2C2C',
    warningLight: '#EAEAEA',
    success: '#334C96',
    successLight: '#d9e0e7',
    disabled: '#C7C7C7',
    typographyDefault: '#2C2C2C',
  },
  space: [0, 4, 8, 12, 16, 24, 32, 48, 64, 96, 128, 144, 192, 256],
  borders: [0, '1px solid', '2px solid', '3px solid'],
  breakpoints: ['768px', '992px', '1200px'],
  fontWeights: {
    regular: 'normal',
    bold: 700,
  },
  fontSizes: {
    sm: '12px',
  },
  shadows: {
    smooth: {
      md: '0.1px 0.3px 0.3px rgba(0, 0, 0, 0.02),0.3px 0.7px 0.9px rgba(0, 0, 0, 0.028),0.7px 1.4px 1.8px rgba(0, 0, 0, 0.035),1.5px 2.9px 3.7px rgba(0, 0, 0, 0.047),4px 8px 10px rgba(0, 0, 0, 0.07)',
    },
    box: {
      subNav: '0px 2px 4px rgba(0, 0, 0, 0.1)',
    },
  },
  zIndices: {
    accordionItemTitle: 200,
    modal: 13001,
    sectionNav: 13000,
  },
  transitionDuration: {
    t1: '1s',
  },
  transitionTimingFunction: {
    default: 'ease-in-out',
    out: 'ease-out',
    in: 'ease-in',
  },
  name: HONDA_THEME_NAME,
  fonts: {
    default: 'Helvetica',
    bold: 'Cooper Black',
    heading: 'Roboto',
  },
  header: {
    desktop: {},
    mobile: {},
  },
  icons: {},
  logIconWarnings: false,
};

export default mockTheme;
