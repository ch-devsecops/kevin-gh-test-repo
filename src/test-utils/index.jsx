import React from 'react';
import { render } from '@testing-library/react';
import { ThemeProvider } from 'styled-components';
import { SitecoreContext } from '@sitecore-jss/sitecore-jss-react';
import { BrowserRouter } from 'react-router-dom';
import { OptionalVideoModalContextProvider } from '@honda-canada/design-system-react';
import UserLocationProvider from '@honda-canada/user-location';
import { configureAxe } from 'jest-axe';
import ModelExplorationContextProvider from '../components/ModelExplorationContext';
import ModelFiltersContextProvider from '../components/ModelFiltersContext';
import { MockFields, MockRendering } from './mockFields';
import MockTheme from './mockTheme';
import 'intersection-observer';
import { ACURA_THEME_NAME } from '../utils/constants';

const generateJssWrapper = context =>
  // eslint-disable-next-line react/display-name
  function ({ children }) {
    return (
      <SitecoreContext
        context={{
          language: 'en',
          site: { name: ACURA_THEME_NAME },
          ...context?.sitecoreContext,
        }}
        componentFactory={{}}
      >
        <BrowserRouter>
          <UserLocationProvider>
            <ThemeProvider theme={context?.theme || MockTheme}>
              <OptionalVideoModalContextProvider>
                <ModelExplorationContextProvider>
                  <ModelFiltersContextProvider>{children}</ModelFiltersContextProvider>
                </ModelExplorationContextProvider>
              </OptionalVideoModalContextProvider>
            </ThemeProvider>
          </UserLocationProvider>
        </BrowserRouter>
      </SitecoreContext>
    );
  };
/**
 * A custom implementation of the testing-library `render` function that
 * wraps tests in an typical JSS app's context providers.
 * Note that you can include a `context` object to set or override values
 * for SitecoreContext and ThemeContext.
 * @param {ReactElement} component - the component you want testing-library to render
 * @param {object} [options] - see https://testing-library.com/docs/react-testing-library/api/#render-options
 * @param {string} [options.context] - default values for SitecoreContext, ThemeProvider
 * @param {string} [options.context.sitecoreContext] - value for SitecoreContext
 * @param {string} [options.context.theme] - value for ThemeProvider
 * @example
 * const { getByText } = jssRender(
 *  <MyComponent />, { context: { sitecoreContext: { language: 'fr' } } }
 * );
 * @returns {function(): void} custom testing-library render function
 */
const jssRender = (component, options = {}) => {
  const { context } = options;
  // eslint-disable-next-line no-param-reassign
  delete options.context;

  return render(component, {
    wrapper: generateJssWrapper(context),
    ...options,
  });
};

const generateWrapper = context =>
  // eslint-disable-next-line react/display-name
  function ({ children }) {
    return (
      <BrowserRouter>
        <ThemeProvider theme={context?.theme || MockTheme}>
          <OptionalVideoModalContextProvider>{children}</OptionalVideoModalContextProvider>
        </ThemeProvider>
      </BrowserRouter>
    );
  };

/**
 * A custom implementation of the testing-library `render` function that
 * wraps tests in typical design-system context providers.
 * Note that you can include a `theme` object to set or override values
 * for ThemeContext.
 * @param {ReactElement} component - the component you want testing-library to render
 * @param {object} [options] - see https://testing-library.com/docs/react-testing-library/api/#render-options
 * @param {string} [options.theme] - default value for ThemeProvider
 * @example
 * const { getByText } = render(
 *  <MyComponent />, { theme: { space: { wrapperGap: '15px' } } }
 * );
 * @returns {function(): void} custom testing-library render function
 */
const customRender = (component, options = {}) => {
  const { theme } = options;
  // eslint-disable-next-line no-param-reassign
  delete options.theme;

  return render(component, {
    wrapper: generateWrapper(theme),
    ...options,
  });
};

// disable accessibility checks for errors that are usually trigged by content mistakes
const axe = configureAxe({
  rules: {
    'duplicate-id': { enabled: false },
    'heading-order': { enabled: false },
  },
});

export * from 'jest-axe';
export { axe };
export * from '@testing-library/react';
export { customRender as render, jssRender };
export { MockFields, MockRendering, MockTheme };
