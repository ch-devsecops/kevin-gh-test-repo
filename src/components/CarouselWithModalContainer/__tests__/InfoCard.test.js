import React from 'react';
import { jssRender, cleanup, axe } from 'test-utils';
import InfoCard from '../InfoCard';

afterEach(cleanup);

const mockData = require('../../../../sample-mock-data/routes/carousel-with-modal-container/en.json');

const mockFields = mockData?.fields?.items[0];

describe('<InfoCard>', () => {
  it('should render correctly', () => {
    const { container } = jssRender(<InfoCard fields={mockFields?.fields} />);

    expect(container).toMatchSnapshot();
    expect(container.children[0]).toBeTruthy();
  });

  it('should return null if there is no fields', () => {
    const { container } = jssRender(<InfoCard fields={null} />);

    expect(container).toMatchSnapshot();
    expect(container.firstChild).toBeNull();
  });

  it('should have no accessibility violations', async () => {
    const { container } = jssRender(<InfoCard fields={mockFields?.fields} />);

    expect(container).toMatchSnapshot();
    expect(await axe(container)).toHaveNoViolations();
  });
});
