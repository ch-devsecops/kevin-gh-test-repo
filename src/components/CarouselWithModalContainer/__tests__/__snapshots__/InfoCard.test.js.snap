// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`<InfoCard> should have no accessibility violations 1`] = `
<div>
  <div
    class="Box-sc-1bsf784-0 sc-1k44dfv-0 dLNGOD"
    height="100%"
  >
    <div
      class="Box-sc-1bsf784-0 sc-1k44dfv-0 hiduYi"
      color="black"
      data-gtm-category="Others"
    >
      <div
        class="Box-sc-1bsf784-0 iPRsIc"
      >
        <img
          alt="Image"
          class="sc-1k44dfv-0 jLTFZk"
          draggable="true"
          loading="lazy"
          src="./sample-mock-data/media/img/placeholders/placeholder-landscape.png"
        />
      </div>
      <div
        class="Box-sc-1bsf784-0 sc-1k44dfv-0 kJntAi"
        color="black"
      >
        <div
          class="sc-1k44dfv-0 jcrtwT sc-1k44dfv-0 jkQLKZ"
          color="inherit"
        >
          

          <h6
            class="sc-1k44dfv-0 eeeuqx"
          >
            01 - Automatic mechanical de-compression system (Main title)
          </h6>
        </div>
        <div
          class="sc-1k44dfv-0 jcrtwT sc-1k44dfv-0 bSotsc"
          color="inherit"
        >
          

          <p>
            Lorem ipsum dolor sit amet, consectetur adipiscing elit. Aenean finibus enim eu enim cursus, sollicitudin sollicitudin nibh euismod. Praesent lobortis ipsum ligula. Donec diam lectus, consectetur ac lectus eu, cursus consequat lectus. Lorem ipsum dolor sit amet, consectetur adipiscing elit.
          </p>
        </div>
        <div
          class="Box-sc-1bsf784-0 jKsJVt"
        >
          <button
            class="Link-sc-16wd4a0-0 dIBUJD"
            color="black !important"
          >
            Shared.Common.learnMoreLabel
          </button>
        </div>
      </div>
    </div>
  </div>
</div>
`;

exports[`<InfoCard> should render correctly 1`] = `
<div>
  <div
    class="Box-sc-1bsf784-0 sc-1k44dfv-0 dLNGOD"
    height="100%"
  >
    <div
      class="Box-sc-1bsf784-0 sc-1k44dfv-0 hiduYi"
      color="black"
      data-gtm-category="Others"
    >
      <div
        class="Box-sc-1bsf784-0 iPRsIc"
      >
        <img
          alt="Image"
          class="sc-1k44dfv-0 jLTFZk"
          draggable="true"
          loading="lazy"
          src="./sample-mock-data/media/img/placeholders/placeholder-landscape.png"
        />
      </div>
      <div
        class="Box-sc-1bsf784-0 sc-1k44dfv-0 kJntAi"
        color="black"
      >
        <div
          class="sc-1k44dfv-0 jcrtwT sc-1k44dfv-0 jkQLKZ"
          color="inherit"
        >
          

          <h6
            class="sc-1k44dfv-0 eeeuqx"
          >
            01 - Automatic mechanical de-compression system (Main title)
          </h6>
        </div>
        <div
          class="sc-1k44dfv-0 jcrtwT sc-1k44dfv-0 bSotsc"
          color="inherit"
        >
          

          <p>
            Lorem ipsum dolor sit amet, consectetur adipiscing elit. Aenean finibus enim eu enim cursus, sollicitudin sollicitudin nibh euismod. Praesent lobortis ipsum ligula. Donec diam lectus, consectetur ac lectus eu, cursus consequat lectus. Lorem ipsum dolor sit amet, consectetur adipiscing elit.
          </p>
        </div>
        <div
          class="Box-sc-1bsf784-0 jKsJVt"
        >
          <button
            class="Link-sc-16wd4a0-0 dIBUJD"
            color="black !important"
          >
            Shared.Common.learnMoreLabel
          </button>
        </div>
      </div>
    </div>
  </div>
</div>
`;

exports[`<InfoCard> should return null if there is no fields 1`] = `<div />`;
